### ISSUE TITLE
Faulty? mt942 from mBank

### ISSUE THREAD

**Issue by sq9mev** [WoLpH/mt940#31]
I'ma trying to parse mt942 provided by polish mBank, which throws:

```
  File "/home/users/bociek/lmsmap/eggs/mt_940-4.3-py2.7.egg/mt940/tags.py", line 70, in parse
    self, value)
RuntimeError: ('Unable to parse "<mt940.tags.FloorLimitIndicator object at 0x7fab98295050>" from "PLN0"', <mt940.tags.FloorLimitIndicator object at 0x7fab98295050>, 'PLN0')

```

That is because of line:

```
:34F:PLN0
```

Which does not match https://github.com/WoLpH/mt940/blob/develop/mt940/tags.py#L162 regex - the line is lacking credit/debit mark.

I'm not sure if credit/debit mark is obligatory here, however there's one thing i know almost for sure - mBank will not fix it ;).

i can imagine two resolutions:

1. make regrex [DC] optional, if it's optional indeed
2. add support for another preparse_ processors, which could be run somewhere before https://github.com/WoLpH/mt940/blob/develop/mt940/models.py#L341

I need advice what is the better solutoin, anybody?

**Comment by sq9mev** [comment]
Well, seems like debit mark is optional: https://www2.swift.com/uhbonline/books/public/en_uk/us9m_20160722/htframe.htm

but it not quite clear for me:

> When D/C Mark is present, it must contain the following code (Error code(s): T51):
> D	Debit	Debit floor limit.

So, only "D" is allowed??

and:

> PRESENCE
> Mandatory
> …
> When field 34F Credit Floor Limit Indicator is absent, the floor limit applies to both debit and credit amounts. When different limits apply, field 34F Credit Floor Limit Indicator must be present.

How it's possible than presence od 34F is mandatory and last sentence allows it's absence? It' contradictory, isn't it?

**Comment by WoLpH** [comment]
As I understand it (from your link) the pattern is: `3!a[1!a]15d`
Which means:

 - `3!a`: 3 required characters (the currency in this case)
 - `[1!a]` an optional character for debit/credit (C or D)
 - `15d` a 15 character number decimal number

It seems when I accepted the pull request for the MT942 support (which added the 34F statement) I didn't check the code thoroughly enough. The pattern seems to be broken: https://github.com/WoLpH/mt940/blob/develop/mt940/tags.py#L162

Instead of `(?P<status>[A-Z]?[DC])  # 2a Debit/Credit Mark` that should be: `(?P<status>[DC]?)  # 2a Debit/Credit Mark` I think.

Can you try the current development release? I think my last change might fix the issue.

**Comment by sq9mev** [comment]
Thanks @WoLpH,  I've forked this repo, and working on https://github.com/sq9mev/mt940/tree/feature/mbank.

Investigating further i've found some more issues with mt942:

- currency of transactions can not be recognized correctly
- sum_credit_debit_entries adds the same keys to transactions.data overwriting each other

You can take a look at my work, i'll try to provide PR with some tests tomorrow.

**Comment by WoLpH** [comment]
Great! Thanks so much for the help :)

**Comment by sq9mev** [comment]
Thanks @WoLpH, closed in #32
